kerastuner/engine/oracle.py

```diff
@@ -17,90 +17,241 @@
+            trial_id=trial_id,
+            status=status)
 
-    def report_status(self, trial_id, status, score=None, t=None):
-        """Used by a worker to report the current status of a trial.
+        if status == trial_lib.TrialStatus.RUNNING:
+            self.ongoing_trials[tuner_id] = trial
+            self.trials[trial_id] = trial
+
+        return trial
+
+    def update_trial(self, trial_id, metrics, t=0):
+        """Used by a worker to report the status of a trial.
 
         Args:
             trial_id: A previously seen trial id.
-            status: String, one of "RUNNING", "CANCELLED". A status of
-                "CANCELLED" means a trial has crashed or been deemed
-                infeasible.
-            score: (Optional) Float. The current, intermediate value of the
-                trial's objective.
-            t: (Optional) Float. The current value in a timeseries representing
-                the state of the trial. This is the value that `score` will be
-                associated with.
+            metrics: Dict of float. The current value of this
+                trial's metrics.
+            t: (Optional) Float. Used to report intermediate results. The
+                current value in a timeseries representing the state of the
+                trial. This is the value that `metrics` will be associated with.
 
         Returns:
-            `OracleResponse.STOP` if the trial should be stopped, otherwise 
-            `OracleResponse.OK`.
+            Trial object. Trial.status will be set to "STOPPED" if the Trial
+            should be stopped early.
         """
-        return OracleResponse.OK
+        trial = self.trials[trial_id]
+        self._check_objective_found(metrics)
+        for metric_name, metric_value in metrics.items():
+            trial.metrics.update(metric_name, metric_value, t=t)
+        # To handle early stopping, set Trial.status to "STOPPED".
+        return trial
 
-    def save(self, fname):
-        raise NotImplementedError
+    def end_trial(self, trial_id, status):
+        """Record the measured objective for a set of parameter values.
 
-    def reload(self, fname):
-        raise NotImplementedError
+        Args:
+            trial_id: String. Unique id for this trial.
+            status: String, one of "COMPLETED", "INVALID". A status of
+                "INVALID" means a trial has crashed or been deemed
+                infeasible.
+        """
+        trial = None
+        for tuner_id, ongoing_trial in self.ongoing_trials.items():
+            if ongoing_trial.trial_id == trial_id:
+                trial = self.ongoing_trials.pop(tuner_id)
+                break
+
+        if not trial:
+            raise ValueError(
+                'Ongoing trial with id: {} not found.'.format(trial_id))
+
+        trial.status = status
+        if status == trial_lib.TrialStatus.COMPLETED:
+            trial.score = self.score_trial(trial)
+
+    def score_trial(self, trial):
```

Comment:
This assumes that users always want the best measurement as the trial value. From my experience, this is not true.
Some users want last measurement, and some users may wish to complete the trial with a new measurement.

Reply:
This is why this is a method, so Oracle subclasses can override this. Oracles that want to expose different scoring strategies to users can have an argument that controls this behavior